File: src/Application/Auth/ResetPasswordFailure.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using po_prostu_silka.Application.Members;
using po_prostu_silka.Application.Notifications;
using po_prostu_silka.Application.Persistence;
using po_prostu_silka.Domain;
using po_prostu_silka.Domain.Members;

namespace po_prostu_silka.Application.Auth;

/// <summary>
/// Why the reset failed.
///
/// <c>invalid_token</c> deliberately covers an unknown address, a malformed token, a token belonging
/// to someone else, an already-used token AND an expired one. Splitting those apart would hand an
/// anonymous caller the account-enumeration oracle that <c>/forgot-password</c> is built to deny -
/// "expired" means the address exists. One code, and the screen says "poproś o nowy link".
/// </summary>
public record ResetPasswordFailure(string Reason);

File: src/app/src/app/core/auth/reset-password-failure.ts
import { createFailureMessages } from '../http/failure-messages';
import { ResetPasswordFailureReason } from './auth.models';
import { MIN_PASSWORD_LENGTH } from './validation';

/**
 * What to tell someone whose password reset was refused.
 *
 * `invalid_token` deliberately covers an unknown address, a malformed token, one belonging to
 * somebody else, an already-used one AND an expired one. The API refuses to distinguish them, so
 * this must not imply it can — "wygasł" would tell an anonymous caller the address is registered.
 * The screen does not put it on a control either: no field the member can edit would fix a spent
 * link, and the only way forward is a new one.
 */
const MESSAGES: Record<ResetPasswordFailureReason, string> = {
  invalid_token:
    'Ten link do zmiany hasła jest nieprawidłowy lub został już użyty. Poproś o nowy link.',
  invalid_new_password: `Hasło musi mieć co najmniej ${MIN_PASSWORD_LENGTH} znaków.`,
};

/** The fallback for a refusal with no reason, or one this build does not know. */
const UNKNOWN = 'Nie udało się ustawić nowego hasła. Spróbuj ponownie za chwilę.';

/** The message for a password-reset refusal. */
export const resetPasswordFailureMessage = createFailureMessages(MESSAGES, UNKNOWN);
